lib/query-interface.js

```diff
@@ -190,728 +172,737 @@ module.exports = (function() {
+};
+
+QueryInterface.prototype.getForeignKeysForTables = function(tableNames, options) {
+  var self = this;
+  options = options || {};
+
+  if (tableNames.length === 0) {
+    return Promise.resolve({});
+  }
+
+  return Promise.map(tableNames, function(tableName) {
+    return self.sequelize.query(self.QueryGenerator.getForeignKeysQuery(tableName, self.sequelize.config.database), {logging: options.logging}).get(0);
+  }).then(function(results) {
+    var result = {};
+
+    tableNames.forEach(function(tableName, i) {
+      if (Utils._.isObject(tableName)) {
+        tableName = tableName.schema + '.' + tableName.tableName;
+      }
 
-  QueryInterface.prototype.showIndex = function(tableName, options) {
-    var sql = this.QueryGenerator.showIndexesQuery(tableName, options);
-    options = options || {};
-    return this.sequelize.query(sql, {
-      logging: options.logging,
-      type: QueryTypes.SHOWINDEXES
+      result[tableName] = Utils._.compact(results[i]).map(function(r) {
+        return r.constraint_name;
+      });
     });
-  };
-
-  QueryInterface.prototype.nameIndexes = function(indexes, rawTablename) {
-    return this.QueryGenerator.nameIndexes(indexes, rawTablename);
-  };
-
-  QueryInterface.prototype.getForeignKeysForTables = function(tableNames, options) {
-    var self = this;
-    options = options || {};
 
-    if (tableNames.length === 0) {
-      return Promise.resolve({});
-    }
-
-    return Promise.map(tableNames, function(tableName) {
-      return self.sequelize.query(self.QueryGenerator.getForeignKeysQuery(tableName, self.sequelize.config.database), {logging: options.logging}).get(0);
-    }).then(function(results) {
-      var result = {};
-
-      tableNames.forEach(function(tableName, i) {
-        if (Utils._.isObject(tableName)) {
-          tableName = tableName.schema + '.' + tableName.tableName;
+    return result;
+  });
+};
+
+QueryInterface.prototype.removeIndex = function(tableName, indexNameOrAttributes, options) {
+  options = options || {};
+  var sql = this.QueryGenerator.removeIndexQuery(tableName, indexNameOrAttributes);
+  return this.sequelize.query(sql, {logging: options.logging});
+};
+
+QueryInterface.prototype.insert = function(instance, tableName, values, options) {
+  var sql = this.QueryGenerator.insertQuery(tableName, values, instance && instance.Model.rawAttributes, options);
+
+  options.type = QueryTypes.INSERT;
+  options.instance = instance;
+  return this.sequelize.query(sql, options).then(function(result) {
+    if (instance) result.isNewRecord = false;
+    return result;
+  });
+};
+
+QueryInterface.prototype.upsert = function(tableName, values, model, options) {
+  var wheres = []
+    , where
+    , indexFields
+    , indexes = []
+    , updateValues
+    , attributes = Object.keys(values);
+
+  where = {};
+  for (var i = 0; i < model.primaryKeyAttributes.length; i++) {
+    var key = model.primaryKeyAttributes[i];
+    if (values[key]){
```

Comment:
That would probably work :)
Although i'm not sure it matters after all since this is for primary key attributes which generally can't be null i assume.